### ISSUE TITLE
Bulk loading fails with QFCounttable

### ISSUE THREAD

**Issue by standage** [dib-lab/khmer#1751]
I'm getting the following error message on Mac OS X.

```
>>> import khmer
>>> ct = khmer.QFCounttable(31, 2**16)
>>> ct.consume_seqfile('kevlar/notebook/human-sim-pico/father-reads-dump.fq.gz')
Assertion failed: (start_index <= empty_index && empty_index < qf->xnslots), function shift_remainders, file gqf.c, line 463.
Abort trap: 6
```

**Comment by betatim** [comment]
I think this is the "QF is too full to continue" error message. We discussed this in https://github.com/dib-lab/khmer/pull/1675#issuecomment-310657888 and follow ups. You can keep filling a BF beyond reasonable limits and it won't ever crash. The QF data structure thinks that you should stop. The recommended course of action is to grow the QF, which is reasonable unless you are already using all the RAM you can.

Mostly a question of "better error message needed" and a good idea what to tell the user to do beyond "we are full, buy more RAM"

**Comment by betatim** [comment]
I think this is the relevant comment https://github.com/splatlab/cqf/issues/4#issuecomment-310457762 for figuring out whether or not the QF is about to be full.

**Comment by standage** [comment]
The thing is, the error message happens immediately, I might even guess on the first attempt to insert. I'll be honest and say I don't even know how big the table I've requested is (hence #1748).

**Comment by standage** [comment]
Ok, the example above handles a Fastq file with 5 reads just fine. It still fails on a 13 Mb gzipped Fastq file, which is much bigger but of course orders of magnitude smaller than real Fastq files out in the wild.

I can increase the starting memory for a while, until I run into this error message.

```
>>> ct = khmer.QFCounttable(31, 2**32)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "khmer/_oxli/graphs.pyx", line 255, in khmer._oxli.graphs.QFCounttable.__cinit__ (khmer/_oxli/graphs.cpp:5067)
    def __cinit__(self, int k, int starting_size):
OverflowError: value too large to convert to int
```

We probably want to define this as an `unsigned long int`.

**Comment by standage** [comment]
I modified the `.pyx` file and changed `int starting_size` to `unsigned long int starting_size`. This allows me to request larger starting sizes. But when I run the code above with starting size of `2**6` the process consumes > 10Gb of memory (paging to the SSD on my laptop) and then *still* fails with the Assertion error.

**Comment by ctb** [comment]
might be worth looking into exactly how squeakr does it.

**Comment by standage** [comment]
Yeah, you're right. Sigh...

![afraid-of-that](https://user-images.githubusercontent.com/566823/28883605-e1c54dce-7763-11e7-8b97-def46ff2f98d.gif)

**Comment by betatim** [comment]
Some thoughts: the CQF has a fixed number of slots, which is determined by the starting size parameter. When the first kmer is added it uses one slot, add it again -> use 2 slots, add again -> 3 slots, [after that](https://github.com/splatlab/cqf/issues/4#issuecomment-310453682) 2 + log(C)/r.

The number of slots is 2**qbits or `start_size`.

So you can fill up a CF by repeatedly adding the same kmer.

I'm using the below code to figure things out. Add it to the cqf directory and build after adding a line like `tt:                  tt.o 								 gqf.o` to `Makefile`.

```c++
#include <stdio.h>


#include "gqf.h"

// number of unique kmers being added to the CF
#define N 8

int main(int argc, char **argv) {
  QF cf;
  uint64_t qbits = 3;
  // not sure what this does, seems like you can't set it to anything but
  // qbits + 8
  uint64_t nhashbits = qbits + 8;
  uint64_t nslots = (1ULL << qbits);

  /* Initialise the CQF */
  qf_init(&cf, nslots, nhashbits, 0);

  for (int i=0; i<8000; ++i) {
    qf_insert(&cf, (i%N)%cf.range, 0, 1);
    printf("%i %i xslots:%llu nocc:%llu nunique:%llu count:%llu\n",
           i, i%N, cf.xnslots, cf.noccupied_slots, cf.ndistinct_elts,
           qf_count_key_value(&cf, i%N, 0));
  }
  return 0;
}
```

This crashes when you need to add a 37th slot to keep counting up. I think.

**Comment by standage** [comment]
Oh. So "filling up" is influenced not only by number of distinct k-mers but also by k-mer abundance?

**Comment by betatim** [comment]
I think  `nhashbits = qbits + 8;` because `#define BITS_PER_SLOT 8` in `gqf.h`. If you set that to a bigger number (say 16) you need to increase `nhashbits` accordingly but you can count further in a block before you need to spill over into the next one.

**Comment by betatim** [comment]
Some notes to self while digging around in this:

```python
read = 'CAGCACAGAGGATATCGCTGTTACANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'

def kmers(seq, k=31):
    for i in range(0, len(seq)-k+1):
        yield seq[i:i+k]

ct = khmer.QFCounttable(31, 2**16)
unique_kmers = set()
N=0
for kmer in kmers(read):
    print(N, len(unique_kmers), ct.n_occupied(), ct.n_unique_kmers())
    unique_kmers.add(kmer);ct.add(kmer)
    N+=1
```
first few lines of output:
```
0 0 0 0
1 1 1 1
2 2 2 2
3 3 3 3
4 4 4 4
5 5 5 5
6 6 6 6
7 7 7 7
8 8 8 8
9 9 9 8
10 10 10 9
11 11 11 10
12 12 12 10
...
```

Not sure I grok why this happens.

Running squeaker on a fastq file (after changing `K` to 31 in `kmer.h`)
```
@EAS20_8_6_1_2_768/1
CAGCACAGAGGATATCGCTGTTACANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+
HGHIHHHGHECHHHHHHHGGHHHHH###########################################################################
```
gives:
```
$ ./main 0 16 1 ../khmer/ecoli_ref.first-read.fastq
Reading from the fastq file and inserting in the QF
Total Time Elapsed: 0.000639seconds
Calc freq distribution: 
Total Time Elapsed: 0.000000seconds
Maximum freq: 1
Num distinct elem: 1
Total num elems: 1
```

Overall status: confused.com

**Comment by betatim** [comment]
Mystery one solved: squeaker skips `N`

**Comment by betatim** [comment]
Next puzzle piece: don't use the reversible hash but instead mumur. This is probably because the reversible hash is not at all "random" in how it selects where to insert, which somehow means the QFCounttable thinks it is "full" much earlier. With murmur hash still crashes on a large fastq file but only after the number of occupied slots reaches 68092

**Comment by standage** [comment]
> With murmur hash still crashes on a large fastq file but only after the number of occupied slots reaches 68092.

Great. This is still orders of magnitudes fewer slots than you can do anything useful with in large genomes or metagenomes. I wonder what we're missing...

**Comment by betatim** [comment]
`QFCounttable(31, 2**16)` is quite small no? Can you do something useful with this when using a bloom filter?

**Comment by standage** [comment]
> Can you do something useful with this when using a bloom filter?

No, but my experience was that I could increase the memory steadily until it is beyond what we should reasonably expect but that it would still barf due to saturation.

**Comment by standage** [comment]
Fixed by #1789.
